src/category_theory/sites/dense_subsite.lean

```diff
@@ -0,0 +1,388 @@
+then given any sheaves `ℱ, ℱ'` on `D`, and a morphism `α : G ⋙ ℱ ⟶ G ⋙ ℱ'`, we may glue them
+together to obtain a morphsim of sheaves `ℱ ⟶ ℱ'`.
+- `category_theory.cover_dense.sheaf_iso`: If the `α` above is iso, then the result is also iso.
+- `category_theory.cover_dense.iso_of_restrict_iso`: If `G : C ⥤ (D, K)` is full and cover-dense,
+then given any sheaves `ℱ, ℱ'` on `D`, and a morphism `α : ℱ ⟶ ℱ'`, then `α` is an iso if
+`G ⋙ ℱ ⟶ G ⋙ ℱ'` is iso.
+
+## References
+
+* [Elephant]: *Sketches of an Elephant*, ℱ. T. Johnstone: C2.2.
+* https://ncatlab.org/nlab/show/dense+sub-site
+* https://ncatlab.org/nlab/show/comparison+lemma
+
+-/
+
+universes v
+
+namespace category_theory
+
+variables {C : Type*} [category C] {D : Type*} [category D] {E : Type*} [category E]
+variables (J : grothendieck_topology C) (K : grothendieck_topology D)
+variables {L : grothendieck_topology E}
+
+/--
+Given a functor `G`, the presieve of `U : D` such that each arrows factors through images of `G`.
+-/
+def presieve.cover_by_image (G : C ⥤ D) (U : D) : presieve U :=
+λ Y f, ∃ (Z : C) (f₁ : Y ⟶ G.obj Z) (f₂ : G.obj Z ⟶ U), f = f₁ ≫ f₂
+
+/--
+An auxiliary structure that witnesses the fact that `f` factors through an image object of `G`.
+-/
+@[nolint has_inhabited_instance]
+structure presieve.cover_by_image_structure (G : C ⥤ D) {V U : D} (f : V ⟶ U) :=
+(obj : C) (lift : V ⟶ G.obj obj) (map : G.obj obj ⟶ U) (fac' : f = lift ≫ map)
+
+@[simp, reassoc]
+lemma presieve.cover_by_image_structure.fac {G : C ⥤ D} {V U : D} {f : V ⟶ U}
+  (h : presieve.cover_by_image_structure G f) : h.lift ≫ h.map = f := h.fac'.symm
+
+/-- Obtain the auxiliary structure given `f ∈ cover_by_image G U`. -/
+noncomputable
+def presieve.get_cover_by_image_structure {G : C ⥤ D} {V U : D} {f : V ⟶ U}
+  (hf : presieve.cover_by_image G U f) : presieve.cover_by_image_structure G f :=
+begin
+  choose a b c d using hf,
+  exact ⟨a, b, c, d⟩,
+end
+
+/--
+Given a functor `G`, the sieve of `U : D` such that each arrows factors through images of `G`.
+-/
+def sieve.cover_by_image (G : C ⥤ D) (U : D) : sieve U :=
+⟨presieve.cover_by_image G U,
+  λ X Y f ⟨Z, f₁, f₂, eq⟩ g, ⟨Z, g ≫ f₁, f₂, by rw [category.assoc, ←eq] ⟩⟩
+
+lemma presieve.in_cover_by_image (G : C ⥤ D) {X : D} {Y : C} (f : G.obj Y ⟶ X) :
+  presieve.cover_by_image G X f := ⟨Y, 𝟙 _, f, by simp⟩
+
+/--
+A functor `G : (C, J) ⥤ (D, K)` is called `cover_dense` if for each object in `D`,
+  there exists a covering sieve in `D` that factors through images of `G`.
+
+This definition can be found in https://ncatlab.org/nlab/show/dense+sub-site Definition 2.2.
+-/
+@[nolint has_inhabited_instance]
+structure cover_dense (K : grothendieck_topology D) (G : C ⥤ D) :=
+(is_cover : ∀ (U : D), sieve.cover_by_image G U ∈ K U)
+
+open presieve opposite
+namespace cover_dense
+variables {A : Type*} [category A] {K} {G : C ⥤ D} (H : cover_dense K G)
+
+lemma sheaf_ext (H : cover_dense K G) (ℱ : SheafOfTypes K) (X : D) {s t : ℱ.val.obj (op X)}
```

Comment:
I think this name is a bit misleading, you're not showing that too sheaves are equal but rather that two elements of one sheaf are equal